pkg/manifest/manifest.go

```diff
@@ -98,6 +105,100 @@ func (m *Manifest) UnmarshalJSON(in []byte) error {
 	return validateResourceId(m.id)
 }
 
+// Include returns an error if the manifest fails an inclusion filter and should not be excluded from futher
+// processing by cluster version operator. Pointer arguments can be set nil to avoid excluding based on that
+// filter. For example, setting profile non-nil and capabilities nil will return an error if the manifest's
+// profile does not match, but will never return an error about capability issues.
+func (m *Manifest) Include(excludeIdentifier *string, includeTechPreview bool, profile *string, capabilities *configv1.ClusterVersionCapabilitiesStatus) error {
+
+	annotations := m.Obj.GetAnnotations()
+	if annotations == nil {
+		return fmt.Errorf("no annotations")
+	}
+
+	var excludeAnnotation string
+	if excludeIdentifier != nil {
+		excludeAnnotation = fmt.Sprintf("exclude.release.openshift.io/%s", *excludeIdentifier)
+	}
+	if v := annotations[excludeAnnotation]; v == "true" {
+		return fmt.Errorf("%s=%s", excludeAnnotation, v)
+	}
+
+	featureGateAnnotation := "release.openshift.io/feature-gate"
+	featureGateAnnotationValue, featureGateAnnotationExists := annotations[featureGateAnnotation]
+	if featureGateAnnotationValue == string(configv1.TechPreviewNoUpgrade) && !includeTechPreview {
+		return fmt.Errorf("tech-preview excluded, and %s=%s", featureGateAnnotation, featureGateAnnotationValue)
+	}
+	// never include the manifest if the feature-gate annotation is outside of allowed values (only TechPreviewNoUpgrade is currently allowed)
+	if featureGateAnnotationExists && featureGateAnnotationValue != string(configv1.TechPreviewNoUpgrade) {
+		return fmt.Errorf("unrecognized value %s=%s", featureGateAnnotation, featureGateAnnotationValue)
+	}
+
+	var profileAnnotation string
+	if profile != nil {
+		profileAnnotation = fmt.Sprintf("include.release.openshift.io/%s", *profile)
+	}
+
+	if val, ok := annotations[profileAnnotation]; ok && val != "true" {
+		return fmt.Errorf("unrecognized value %s=%s", profileAnnotation, val)
+	} else if !ok {
+		return fmt.Errorf("%s unset", profileAnnotation)
+	}
+
+	// If there is no capabilities defined in a release then we do not need to check presence of capabilities in the manifest
```

Comment:
This is really "the caller didn't ask us to filter by capability", which is fairly orthogonal to whether there are capabilities defined in the release.  And we already cover the meaning of `nil` arguments in English in the functions godocs, so I expect we can just drop this line.  Also, just a comment, so if you don't want to drop or rephrase, I don't have a problem landing the PR with this line unchanged.

Reply:
It is not big deal , but the comment will help the code readability. Lets keep it.